File: src/clj/et/rec/ocr.clj
(ns et.rec.ocr
  "Reading the text off a frame, with Vision.

   **Why an OCR and not a model that is shown the picture and asked.** The
   terms are known strings — an address, a key, a client's name, typed into a
   box by the person who wants them gone. So this is not a recognition problem
   with a judgement in it; it is a search, and a search should be repeatable,
   inspectable, and the same on Tuesday as it was on Monday. A vision model
   asked to find secrets in a screenshot answers differently each time it is
   asked and cannot say where on the frame it looked. What is wanted here is a
   box with coordinates.

   **Why Vision and not tesseract**, which this workspace already has for
   `rhizome-books`. Measured on a real 2560x1440 frame of a screencast:

       Vision      0.24 s/frame    253 words
       tesseract   0.67 s/frame    171 words

   and on the term that mattered — an address rendered in 12px grey — tesseract
   read `dang@eighttrigrams.net` where Vision read it correctly at confidence
   1.0. Both numbers point the same way, and the second one is the one that
   decides it: a redaction matching against a corrupted string is a redaction
   that does not happen. Tesseract was built for scans of paper, Vision has
   spent its life on screenshots, and this app only ever looks at screenshots.

   Being macOS-only costs nothing that was not already spent. recorda cannot
   leave the host anyway — AVFoundation for the screen, CoreAudio for the
   interface — so a third framework from the same box is not a new constraint.

   The helper itself is `scripts/ocr.swift`, compiled on demand into `target/`
   rather than committed as a binary. A binary in the repo is a thing to
   license, to keep and to explain, and this one takes under two seconds to
   build."
  (:require [clojure.java.io :as io]
            [clojure.java.shell :as shell]
            [clojure.string :as str]
            [taoensso.telemere :as t])
  (:import (java.io File)
           (java.lang ProcessBuilder$Redirect)))

(def ^:private source (io/file "scripts/ocr.swift"))
(def ^:private compiled (io/file "target/ocr"))

(defn available?
  "Whether this machine can do it at all. Only macOS has Vision, and only a
   machine with the command line tools can build the helper — but recorda is
   already host-and-macOS-only for the screen, so the honest failure here is a
   missing Xcode CLT and not a missing platform."
  []
  (and (str/includes? (str/lower-case (System/getProperty "os.name" "")) "mac")
       (.exists ^File source)
       (zero? (:exit (shell/sh "which" "swiftc")))))

(defn binary
  "The compiled helper, built if it is missing or older than its source.

   Throws rather than returning nil, because every caller wants the binary and
   none of them has anything sensible to do without it."
  ^File []
  (when-not (.exists ^File source)
    (throw (ex-info "scripts/ocr.swift is missing" {})))
  (when (or (not (.exists ^File compiled))
            (< (.lastModified ^File compiled) (.lastModified ^File source)))
    (io/make-parents compiled)
    (t/log! :info "recorda: building the OCR helper")
    (let [{:keys [exit err]} (shell/sh "swiftc" "-O" "-o" (str compiled) (str source))]
      (when-not (zero? exit)
        (throw (ex-info (str "could not build the OCR helper: " err) {})))))
  compiled)

;; --- reading the helper's output -------------------------------------------

(defn- parse-word [line]
  ;; Limit 8 so a tab that somehow survived in the text does not eat it. The
  ;; helper already replaces them, and this is the belt to that's braces.
  (let [[_ li x y w h conf text] (str/split line #"\t" 8)]
    {:line (parse-long li)
     :x    (parse-long x)
     :y    (parse-long y)
     :w    (parse-long w)
     :h    (parse-long h)
     :conf (or (parse-double conf) 0.0)
     :text (or text "")}))

(defn- collect
  "Turn the helper's flat rows into one entry per frame, in the order the
   frames were given. A frame with no text still gets an entry — the helper
   emits its `@` either way, which is what lets a clean frame be told apart
   from one that was never looked at."
  [lines on-frame]
  (loop [[l & more] lines, cur nil, acc []]
    (cond
      (nil? l) (if cur (conj acc cur) acc)

      (str/starts-with? l "@\t")
      (do (when cur (on-frame))
          (recur more {:path (subs l 2) :words []} (if cur (conj acc cur) acc)))

      (str/starts-with? l "w\t")
      (recur more (update cur :words conj (parse-word l)) acc)

      :else (recur more cur acc))))

(defn- run-worker
  "One helper process over one list of frames.

   Stdin is written from another thread while stdout is read on this one. A
   long take is thousands of paths, which is more than a pipe buffer holds, and
   writing them all before reading anything is the classic way to deadlock two
   processes that are each waiting for the other to drain."
  [paths on-frame]
  (let [p       (-> (ProcessBuilder. ^java.util.List [(str (binary))])
                    (.redirectError ProcessBuilder$Redirect/INHERIT)
                    (.start))
        writing (future
                  (with-open [w (io/writer (.getOutputStream p))]
                    (doseq [path paths]
                      (.write w (str path "\n")))))
        rows    (with-open [r (io/reader (.getInputStream p))]
                  (doall (collect (line-seq r) on-frame)))]
    @writing
    (.waitFor p)
    ;; The last frame's `@` has no successor to close it, so its callback fires
    ;; here rather than in `collect`.
    (when (seq rows) (on-frame))
    rows))

(def ^:private workers
  "Three, not one per core.

   Vision threads inside itself — one process already runs at about 140% CPU —
   so the processes are competing for the same units past a point. Measured on
   63 frames of a real take: one process 15.1 s, four processes 9.7 s. That is
   1.56x for 4x the processes, and the third one is where the curve flattens."
  3)

(defn read-frames
  "OCR a list of image files. Answers one `{:path :words}` per frame, in order.

   `on-frame` is called once per frame finished, on whichever worker's thread
   got there — it is for a progress counter and nothing that needs ordering."
  [files & [{:keys [on-frame] :or {on-frame (fn [])}}]]
  (let [paths (mapv str files)]
    (if (empty? paths)
      []
      (let [chunks (partition-all (max 1 (long (Math/ceil (/ (count paths)
                                                             (double workers)))))
                                  paths)
            done   (mapv #(future (run-worker % on-frame)) chunks)]
        (into [] (mapcat deref) done)))))

File: scripts/ocr.swift
// Reads image paths on stdin, one per line. Writes what Vision found on stdout.
//
// Why a Swift file in a Clojure repo: the app is macOS-only already — it needs
// AVFoundation for the screen and CoreAudio for the interface, which is why it
// cannot run in the devbox. Vision is the same kind of thing, sitting in the
// same frameworks, and it reads a screenshot far better than the alternative.
// Measured on a real 2560x1440 frame of a screencast: Vision 0.24 s and 253
// words, tesseract 0.67 s and 171 words — and on the one term that mattered,
// an address in 12px grey, tesseract read `dang@eighttrigrams.net` where
// Vision read it correctly. A redaction that matches on a corrupted string is
// a redaction that does not happen.
//
// Paths come in on stdin rather than argv because a long take is thousands of
// frames and argv is not.
//
// Output, tab separated, one frame's rows following its `@`:
//
//   @  <path>
//   w  <line> <x> <y> <w> <h> <confidence> <text>
//
// A frame with no text still gets its `@`, so the caller can tell "nothing
// here" from "not looked at yet" — which is the difference between a clean
// frame and a silently skipped one.

import Foundation
import Vision
import AppKit

// Small text is the whole point: an address in a log line is 12px in a 1440p
// frame, which is 0.008 of the height. Vision's own default is 1/32, and at
// that setting the things worth redacting are exactly the things it skips.
let minHeight = Float(ProcessInfo.processInfo.environment["RECORDA_OCR_MIN_HEIGHT"] ?? "") ?? 0.008
let langs = (ProcessInfo.processInfo.environment["RECORDA_OCR_LANGS"] ?? "en-US")
    .split(separator: ",").map(String.init)

let out = FileHandle.standardOutput

func emit(_ s: String) {
    if let d = s.data(using: .utf8) { out.write(d) }
}

func scan(_ path: String) {
    emit("@\t\(path)\n")
    guard let img = NSImage(contentsOfFile: path),
          let cg = img.cgImage(forProposedRect: nil, context: nil, hints: nil) else {
        FileHandle.standardError.write("recorda-ocr: cannot read \(path)\n".data(using: .utf8)!)
        return
    }
    let W = CGFloat(cg.width), H = CGFloat(cg.height)

    let req = VNRecognizeTextRequest()
    req.recognitionLevel = .accurate
    // Off, deliberately. Language correction turns an API key into a word and
    // an address into prose — it improves reading and ruins matching.
    req.usesLanguageCorrection = false
    req.minimumTextHeight = minHeight
    req.recognitionLanguages = langs

    let handler = VNImageRequestHandler(cgImage: cg, options: [:])
    do { try handler.perform([req]) } catch {
        FileHandle.standardError.write("recorda-ocr: \(path): \(error)\n".data(using: .utf8)!)
        return
    }
    guard let obs = req.results else { return }

    for (li, o) in obs.enumerated() {
        guard let cand = o.topCandidates(1).first else { continue }
        let s = cand.string
        var cursor = s.startIndex
        for word in s.split(separator: " ", omittingEmptySubsequences: true) {
            guard let r = s.range(of: String(word), range: cursor..<s.endIndex) else { continue }
            cursor = r.upperBound
            // Vision can give a box for any character range of a line, which
            // is what makes a word the unit here rather than a whole line: a
            // line is often most of the screen's width, and blurring one to
            // hide an address in it would black out the sentence around it.
            guard let bb = try? cand.boundingBox(for: r) else { continue }
            let b = bb.boundingBox
            // Vision's origin is bottom-left and normalised; ffmpeg's is
            // top-left and in pixels.
            let x = b.minX * W
            let y = (1 - b.maxY) * H
            let w = b.width * W
            let h = b.height * H
            let text = word.replacingOccurrences(of: "\t", with: " ")
            emit("w\t\(li)\t\(Int(x.rounded(.down)))\t\(Int(y.rounded(.down)))\t\(Int(w.rounded(.up)))\t\(Int(h.rounded(.up)))\t\(cand.confidence)\t\(text)\n")
        }
    }
}

while let line = readLine(strippingNewline: true) {
    let p = line.trimmingCharacters(in: .whitespaces)
    if !p.isEmpty { scan(p) }
}
